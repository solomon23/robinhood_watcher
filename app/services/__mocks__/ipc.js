export const saveSettings = jest.fn()
export const quit = jest.fn()
export const logout = jest.fn()
