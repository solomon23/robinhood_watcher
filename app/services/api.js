// @flow
import 'node-fetch'
import Store from 'electron-store'

const API_ROOT = 'https://api.robinhood.com'
const store = new Store()
const headers = {
  Origin: 'electron://robinhood-app',
  'content-type': 'application/json',
}

export function getToken(): { token: ?string, refresh_token: ?string } {
  return store.get('token') || { token: null, refresh_token: null }
}

function setToken(tokens) {
  store.set('token', tokens)
}

export function getSettings(): ?UserSettings {
  return store.get('settings') || null
}

export function setSettings(settings: UserSettings) {
  store.set('settings', settings)
}

function getErroFromJson(json) {
  let message = ''
  if (json.non_field_errors) {
    [message] = json.non_field_errors
  } else if (json.detail) {
    message = json.detail
  } else if (json.error) {
    message = json.error
  }

  return message
}

export function logout() {
  setToken({ token: null, refresh_token: null })
}

async function makeCall(url) {
  const res = await authRequest(`${API_ROOT}${url}`)
  const json = await res.json()

  if (res.ok) {
    return json
  }

  throw res.code
}

export async function getSMA(symbol: string): Promise<any> {
  return (await fetch(`https://api.iextrading.com/1.0/stock/${symbol}/chart/1d`)).json()
}

export async function getAccountNumber() {
  const json = await makeCall('/accounts/')
  return json.results[0].account_number
}

export async function getPortfolio({ user }: { user: User }): Promise<Portfolio> {
  const json = await makeCall(`/accounts/${user.accountNumber || ''}/portfolio/`)
  return json
}

export async function getWatchlist() {
  const json = await makeCall('/watchlists/Default/')

  return Promise.all(
    json.results.map(async result => {
      const instrument = await (await fetch(
        decodeURIComponent(result.instrument)
      )).json()

      return (await fetch(
        decodeURIComponent(instrument.quote)
      )).json()
    })
  )
}

export async function tryAuthenticate(): Promise<{ authenticated: boolean, error: ?string }> {
  try {
    const json = await makeCall('/accounts/')
    if (json && json.results) {
      return { authenticated: true, error: null }
    }

    return { authenticated: false, error: 'Error' }
  } catch(err) {
    return {
      authenticated: false,
      error: `${err}`,
    }
  }
}

export async function getAccountPositions({ user }: { user: User }): Promise<Array<Position>> {
  const positions = await makeCall(`/accounts/${user.accountNumber || ''}/positions/`)
  const transformed = await Promise.all(
    positions.results
      .filter(result => Number(result.quantity) !== 0)
      .map(async result => {
        const instrument = await (await fetch(
          decodeURIComponent(result.instrument)
        )).json()

        const quote = await (await fetch(
          decodeURIComponent(instrument.quote)
        )).json()

        return {
          ...quote,
          averageBuyPrice: Number(result.average_buy_price),
          quantity: Number(result.quantity),
          currentPrice: Number(quote.last_traded_price),
          symbol: instrument.symbol,
          name: instrument.name,
          instrument,
        }
      })
  )

  return transformed
}

async function authRequest(url: string, options: any = {}): Promise<any> {
  const { token } = getToken()

  const opts = {
    ...options,
    Accept: 'application/json',
    headers: {
      ...headers,
      Authorization: `Bearer ${token || ''}`,
    },
  }

  if (!token) {
    return { ok: false, code: 401, json: () => ({ error: 'Missing auth token', code: 401 }) }
  }

  let results = await fetch(url, opts)
  if (results.status === 401) {
    // try to refresh the token
    try {
      const refresh = await refreshAuthenticate()

      if (refresh.token) {
        setToken({ token: refresh.token, refresh_token: refresh.refresh_token })

        // try again
        opts.headers.Authorization = `Bearer ${token}`
        results = await fetch(url, opts)
      } else {
        setToken({ token: null, refresh_token: null })

        return { ok: false, code: 401, json: () => ({ error: 'Could not refresh token', code: 401 }) }
      }
    } catch (err) {
      // refresh failed
      setToken({ token: null, refresh_token: null })

      return { ok: false, code: 401, json: () => ({ error: 'Could not refresh token', code: 401 }) }
    }
  }

  return results
}

export async function refreshAuthenticate() {
  const { refresh_token: refreshToken } = getToken()

  try {
    const res = await fetch(`${API_ROOT}/oauth2/token/`, {
      method: 'POST',
      Accept: 'application/json',
      headers,
      body: JSON.stringify({
        'client_id': 'c82SH0WZOsabOXGP2sxqcj34FxkvfnWRZBKlBjFS',
        'grant_type': 'refresh_token',
        refresh_token: refreshToken,
      }),
    })

    const json = await res.json()
    if (res.ok) {
      return {
        twoFactorAuthRequired: !!json.mfa_required,
        token: json.access_token || null,
        refresh_token: json.refresh_token,
      }
    }

    // Get the error message from RobinHood and re-throw it
    const message = getErroFromJson(json)

    return {
      code: res.status,
      message,
    }
  } catch (error) {
    return {
      code: error.code,
      error: error.message,
    }
  }
}

export async function authenticate(username: string, password: string, mfaCode: ?string): Promise<any> {
  try {
    const body: any = {
      username,
      password,
      'client_id': 'c82SH0WZOsabOXGP2sxqcj34FxkvfnWRZBKlBjFS',
      'expires_in': '86400',
      'grant_type': 'password',
      'scope': 'internal',
    }

    if (mfaCode) {
      body.mfa_code = mfaCode
    }

    const res = await fetch(`${API_ROOT}/oauth2/token/`, {
      method: 'POST',
      Accept: 'application/json',
      headers,
      body: JSON.stringify(body),
    })

    const json = await res.json()
    if (res.ok) {
      setToken({ token: json.access_token, refresh_token: json.refresh_token })

      return {
        twoFactorAuthRequired: !!json.mfa_required,
        authenticated: !!json.access_token,
      }
    }

    // Get the error message from RobinHood and re-throw it
    const message = getErroFromJson(json)

    throw new Error(message)

  } catch (error) {
    return {
      error: error.message,
    }
  }
}
