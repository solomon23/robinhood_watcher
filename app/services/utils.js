const currency = require('currency.js')

const currencyOptions = {
  separator: ',',
  decimal: '.',
  symbol: '$',
  formatWithSymbol: true,
}

export function USD(amount) {
  return currency(amount, currencyOptions).format(true)
}

export default { USD }
