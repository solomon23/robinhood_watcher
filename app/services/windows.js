import { remote } from 'electron'

let stockInfoWindow = null
let settingsWindow = null

export const closeSettingsWindow = () => {
  settingsWindow.destroy()
  settingsWindow = null
}

export const createStockWindow = (symbol) => {
  if (stockInfoWindow !== null) {
    stockInfoWindow.destroy()
  }

  stockInfoWindow = new remote.BrowserWindow({
    width: 825,
    height: 625,
    backgroundColor: '#212025',
    center: false,
    title: symbol,
    resizable: false,
    titleBarStyle: 'hidden',
    show: true,
  })

  stockInfoWindow.loadURL(`file://${__dirname}/app.html?symbol=${symbol}`)

  stockInfoWindow.on('close', () => {
    stockInfoWindow = null
  })
}

export const createSettingsWindow = (settings) => {
  if (settingsWindow !== null) {
    settingsWindow.destroy()
  }

  settingsWindow = new remote.BrowserWindow({
    height: 470,
    width: 250,
    resizable: false,
    backgroundColor: '#212025',
    titleBarStyle: 'hidden',
  })

  settingsWindow.loadURL(`file://${__dirname}/app.html?settings=true`)

  settingsWindow.webContents.on('did-finish-load', () => {
    settingsWindow.webContents.send('USER_SETTINGS', settings)
  })

  settingsWindow.on('close', () => {
    settingsWindow = null
  })
}

export default {
  createStockWindow,
}
