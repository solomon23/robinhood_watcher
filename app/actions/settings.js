// @flow
import { getSettings } from '../services/api'
import * as ipc from '../services/ipc'

export const VIEW_BY = {
  PERCENT: 'PERCENT',
  TOTAL: 'TOTAL',
  INDIVIDUAL: 'INDIVIDUAL',
}

export const SETTINGS_UPDATE = 'SETTINGS_UPDATE'
export function updateSettings(settings: UserSettings) {
  return { type: SETTINGS_UPDATE, payload: settings }
}

export function loadSettings() {
  return (dispatch: Dispatch) => {
    // get the settings from the local store
    const settings = getSettings()

    if (settings) {
      dispatch(updateSettings(settings))
    }
  }
}

export function saveSettings(settings: UserSettings) {
  return (dispatch: Dispatch) => {
    // update it locally
    dispatch(updateSettings(settings))

    // save it over the wire
    ipc.saveSettings(settings)
  }
}

export function quit() {
  return () => {
    ipc.quit()
  }
}

export function logout() {
  return () => {
    ipc.logout()
  }
}
