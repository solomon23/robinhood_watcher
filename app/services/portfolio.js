import { USD } from './utils'

export function calculateEquity(portfolio) {
  const equity = USD(portfolio.extended_hours_equity || portfolio.equity)

  let equityDifference = portfolio.equity - portfolio.adjusted_equity_previous_close
  const percentDifference = `${Math.abs(100 * equityDifference / portfolio.adjusted_equity_previous_close).toFixed(2)}%`
  const sign = equityDifference >= 0 ? '+' : ''

  equityDifference = `${sign}${USD(equityDifference)} (${equityDifference >= 0 ? '+' : '-'}${percentDifference})`

  let afterHoursEquityDifference = null
  if (portfolio.extended_hours_equity) {
    afterHoursEquityDifference = portfolio.extended_hours_equity - portfolio.equity

    const afterHoursSign = afterHoursEquityDifference >= 0 ? '+' : ''
    const afterPercent = `${Math.abs(100 * afterHoursEquityDifference / portfolio.equity).toFixed(2)}%`

    afterHoursEquityDifference = `${afterHoursSign}${USD(afterHoursEquityDifference)} (${afterHoursSign}${afterPercent})`
  }

  return {
    equity,
    equityDifference,
    afterHoursEquityDifference,
  }
}

export default {
  calculateEquity,
}
