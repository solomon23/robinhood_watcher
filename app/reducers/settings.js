// @flow
import * as settingsActions from '../actions/settings'

const defaultState = {
  refreshInterval: 10,
  viewChangeBy: settingsActions.VIEW_BY.TOTAL,
  notifyPercent: 0,
}

export default function(state: any = defaultState, action: Action) {
  switch (action.type) {
    case settingsActions.SETTINGS_UPDATE:
      return { ...state, ...action.payload }

    default:
      return state
  }
}
