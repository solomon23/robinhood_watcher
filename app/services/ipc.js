import { ipcRenderer } from 'electron'
import { closeSettingsWindow } from './windows'
import * as settingsActions from '../actions/settings'
import * as appActions from '../actions/app'
import * as userActions from '../actions/user'
import * as api from './api'

export function listen(store: Store) {
  ipcRenderer.on('USER_SETTINGS', (event, arg) => {
    store.dispatch(settingsActions.updateSettings(arg))
  })

  ipcRenderer.on('QUIT', () => {
    store.dispatch(appActions.appQuit())
  })

  ipcRenderer.on('LOGOUT', () => {
    closeSettingsWindow()
    store.dispatch(userActions.logout())
  })

  ipcRenderer.on('USER_SETTINGS_SAVE', (events, arg) => {
    // save to the disk store
    api.setSettings(arg)

    // update the redux store
    store.dispatch(settingsActions.updateSettings(arg))

    // kill the settings window
    closeSettingsWindow()

    // stop and restart the refresh timer
    store.dispatch(appActions.stopRefreshTimer())
    store.dispatch(appActions.startRefreshTimer())
  })
}

export function setTitle(title) {
  ipcRenderer.send('SET_MENU_TITLE', title)
}

export function saveSettings(settings) {
  ipcRenderer.send('MAIN_WINDOW', { channel: 'USER_SETTINGS_SAVE', message: settings })
}

export function quit() {
  ipcRenderer.send('MAIN_WINDOW', { channel: 'QUIT' })
}

export function logout() {
  ipcRenderer.send('MAIN_WINDOW', { channel: 'LOGOUT' })
}
