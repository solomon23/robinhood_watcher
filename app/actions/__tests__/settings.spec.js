import configureMockStore from 'redux-mock-store'
import fetchMock from 'fetch-mock'
import thunk from 'redux-thunk'
import * as actions from '../settings'
import * as ipc from '../../services/ipc'

const mockStore = configureMockStore([thunk])
jest.mock('../../services/api.js')
jest.mock('../../services/ipc.js')

describe('settings actions', () => {
  afterEach(() => {
    fetchMock.restore()
  })

  it('should fire update settings', () => {
    expect(actions.updateSettings({ refreshInterval: 100 })).toMatchSnapshot()
  })

  it('should fire loadSettings', () => {
    const store = mockStore({ })

    store.dispatch(actions.loadSettings())
    expect(store.getActions()).toMatchSnapshot()
  })

  it('should fire saveSettings', () => {
    const store = mockStore({ })

    store.dispatch(actions.saveSettings({ refreshInterval: 5, viewChangeBy: 'TOTAL', notifyPercent: 10 }))
    expect(store.getActions()).toMatchSnapshot()
    expect(ipc.saveSettings).toHaveBeenCalledWith({ refreshInterval: 5, viewChangeBy: 'TOTAL', notifyPercent: 10 })
  })

  it('should fire quit', () => {
    actions.quit()()
    expect(ipc.quit).toHaveBeenCalledWith()
  })

  it('should fire logout', () => {
    actions.logout()()
    expect(ipc.logout).toHaveBeenCalledWith()
  })
})
