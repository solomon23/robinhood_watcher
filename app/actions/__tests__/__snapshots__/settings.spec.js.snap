// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings actions should fire loadSettings 1`] = `
Array [
  Object {
    "payload": Object {
      "notifyPercent": 100,
      "refreshInterval": 1,
      "viewChangeBy": "TOTAL",
    },
    "type": "SETTINGS_UPDATE",
  },
]
`;

exports[`settings actions should fire saveSettings 1`] = `
Array [
  Object {
    "payload": Object {
      "notifyPercent": 10,
      "refreshInterval": 5,
      "viewChangeBy": "TOTAL",
    },
    "type": "SETTINGS_UPDATE",
  },
]
`;

exports[`settings actions should fire update settings 1`] = `
Object {
  "payload": Object {
    "refreshInterval": 100,
  },
  "type": "SETTINGS_UPDATE",
}
`;
